### PR TITLE
Guest Build System: Add x86_32 toolchain, formalize things  bit, cleanup FEXLinuxTests

### PR DESCRIPTION
**Pull request by skmp**
#### Overview
- Formalized guest 32 bit x86 as "x86_32" and guest 64 bit x86 as "x86_64" in the toolchains / build system
- Added x86_32 toolchain file
- Moved toolchain files to Toolchains/
- Main CMakeLists stores toolchain files in `TOOLCHAIN_FILE_${GuestArch}` vars
- Modified FEXLinuxTests to use x86_32 toolchain instead of adding -m32 manually
- Added support to FEXLinuxTests to only compile some architectures via `// arch: arch1, arch2`, remove the 32 bit only configuration

#### Notes
This might need some variations on the installed packages which I haven't fully identified, see #1889

### CODE REVIEW

**Comment by skmp** [comment]
@Sonicadvance1 Looks like the runners need `gcc-i686-linux-gnu g++-i686-linux-gnu gcc-multilib-i686-linux-gnu g++-multilib-i686-linux-gnu` installed.

File: Toolchains/toolchain_x86_64.cmake [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+set(CMAKE_C_COMPILER x86_64-linux-gnu-gcc)
+set(CMAKE_CXX_COMPILER x86_64-linux-gnu-g++)
+
+set(CMAKE_C_COMPILER_TARGET ${triple})
+set(CMAKE_CXX_COMPILER_TARGET ${triple})
+
+set(CMAKE_FIND_ROOT_PATH /usr/x86_64-linux-gnu)
+
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
```
Comment (neobrain) [inline Toolchains/toolchain_x86_64.cmake:16]: Could you document what exactly we're setting these for?

File: unittests/FEXLinuxTests/tests/CMakeLists.txt [hunk 2]
```diff
@@ -37,26 +37,38 @@ foreach(TEST ${TESTS})
...
+
+  # message("test: ${TEST_NAME} |${GUEST_ARCH}| in [${ARCHES}]")
 
-  add_executable(${BIN_NAME_64} ${TEST})
-  set_target_properties(${BIN_NAME_64} PROPERTIES COMPILE_FLAGS "${APPEND_CXX_FLAGS} -g -O2" LINK_FLAGS "${APPEND_LD_FLAGS}")
-  target_link_libraries(${BIN_NAME_64} ${LIBS})
+  if (${GUEST_ARCH} IN_LIST ARCHES)
+    set(BIN_NAME "${TEST_NAME}.${GUEST_ARCH}")
+
+    # message("Added Target ${BIN_NAME}")
```
Comment (neobrain) [inline unittests/FEXLinuxTests/tests/CMakeLists.txt:67]: Leftover debug message?

File: Toolchains/toolchain_x86_64.cmake [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# This is tuned for Ubuntu + gcc
+# Fun Fact: Ubuntu's i386 actually uses the i686 compiler
```
Comment (neobrain) [inline Toolchains/toolchain_x86_64.cmake:2]: The fun fact only needs to be mentioned in the 32-bit toolchain.

**Review by neobrain — CHANGES_REQUESTED** [review]

File: Toolchains/toolchain_x86_64.cmake (hunk 1 above)
Comment (skmp) [reply Toolchains/toolchain_x86_64.cmake:16]: I just copied them from the cmake tutorial

Comment (skmp) [reply Toolchains/toolchain_x86_64.cmake:16]: I'll look them up

File: unittests/FEXLinuxTests/tests/CMakeLists.txt (hunk 2 above)
Comment (skmp) [reply unittests/FEXLinuxTests/tests/CMakeLists.txt:67]: I let them there (and in a few other places) so we enable if we have mysterious issues

File: unittests/FEXLinuxTests/CMakeLists.txt [hunk 4]
```diff
@@ -47,14 +44,24 @@ function(AddTests Tests BinDirectory BitnessList)
       set(VARIATIONS "${TEST_NAME}:")
     endif()
 
+    # Get test arch overrides
+    set(ARCH_REGEX "//[ ]*arch: ([^\n]+)")
```
[the remark this answers is not in the document]
Comment (skmp) [reply unittests/FEXLinuxTests/CMakeLists.txt:48]: We already do this for several other things, so it's not new practice (or convoluted, imho). This is similar to how the gcc-target-tests are done, on which the infra for these FLT is based on.

The arch list restricts the guest arches the test is built for.

There are other ones as well (libs, test cases, compiler flags).

The goal is to have the tests be self-contained, self-descriptive and concise.

File: unittests/FEXLinuxTests/tests/CMakeLists.txt [hunk 5]
```diff
@@ -37,26 +37,38 @@ foreach(TEST ${TESTS})
...
+  string(REGEX MATCH ${ARCH_REGEX} ARCHES ${TEST_CODE})
+  # if cannot handle multiline variables, so we have to match the line first
+  if(${ARCHES} MATCHES ${ARCH_REGEX})
+    set(ARCHES "${CMAKE_MATCH_1}" )
+    string(REGEX REPLACE " |," ";" ARCHES "${ARCHES}")
+  else()
+    set(ARCHES "${GUEST_ARCH}" "")
+  endif()
+
+  # message("test: ${TEST_NAME} |${GUEST_ARCH}| in [${ARCHES}]")
```
[the remark this answers is not in the document]
Comment (skmp) [reply unittests/FEXLinuxTests/tests/CMakeLists.txt:62]: It is intentional, as is common practice all over the source code, so we don't have to write and re-write those.

File: Toolchains/toolchain_x86_64.cmake (hunk 3 above)
Comment (skmp) [reply Toolchains/toolchain_x86_64.cmake:2]: ha, true

File: Toolchains/toolchain_x86_64.cmake (hunk 1 above)
Comment (skmp) [reply Toolchains/toolchain_x86_64.cmake:23]: Added some rough explanations

**Comment by skmp** [comment]
I got this to work with `lld` and `linux-libc-dev:amd64` installed

**Comment by skmp** [comment]
this /should/ work with `g++-multilib-i686-linux-gnu` and `g++-multilib-x86_64-linux-gnu`

**Comment by Sonicadvance1** [comment]
Packages are installed, this now has other build failure or test failure depending.

**Comment by skmp** [comment]
@Sonicadvance1 `armv8.4` runner seems to be missing `libc6-dev-i386-cross`, is `g++-multilib-i686-linux-gnu` installed?

**Comment by Sonicadvance1** [comment]
> g++-multilib-i686-linux-gnu

It's already installed.
```
ryanh@buildbot-macmini-2:~$ sudo apt-get install libc6-dev-i386-cross
[sudo] password for ryanh:
Reading package lists... Done
Building dependency tree
Reading state information... Done
libc6-dev-i386-cross is already the newest version (2.31-0ubuntu9.9cross1).
The following packages were automatically installed and are no longer required:
  libclang-common-10-dev libclang-cpp10 libclang1-10 libomp-10-dev libomp5-10 llvm-10 llvm-10-dev llvm-10-runtime llvm-10-tools
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 118 not upgraded.
ryanh@buildbot-macmini-2:~$ sudo apt-get install g++-multilib-i686-linux-gnu
Reading package lists... Done
Building dependency tree
Reading state information... Done
g++-multilib-i686-linux-gnu is already the newest version (4:9.3.0-1ubuntu2).
The following packages were automatically installed and are no longer required:
  libclang-common-10-dev libclang-cpp10 libclang1-10 libomp-10-dev libomp5-10 llvm-10 llvm-10-dev llvm-10-runtime llvm-10-tools
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 118 not upgraded.

**Comment by skmp** [comment]
Hmm, that is odd. I guess we have some other dependency i'm not aware of.

**Comment by skmp** [comment]
Oh wait, we are missing teh `x86_64` ones

**Comment by skmp** [comment]
`g++-multilib-x86_64-linux-gnu` and `libc6-dev-amd64-cross` ?

**Comment by Sonicadvance1** [comment]
> Oh wait, we are missing teh `x86_64` ones

```
ryanh@buildbot-macmini-2:~$ sudo apt-get install libc6-dev-amd64-cross  g++-multilib-x86-64-linux-gnu
Reading package lists... Done
Building dependency tree
Reading state information... Done
g++-multilib-x86-64-linux-gnu is already the newest version (4:9.3.0-1ubuntu2).
libc6-dev-amd64-cross is already the newest version (2.31-0ubuntu9.9cross1).
The following packages were automatically installed and are no longer required:
  libclang-common-10-dev libclang-cpp10 libclang1-10 libomp-10-dev libomp5-10 llvm-10 llvm-10-dev llvm-10-runtime llvm-10-tools
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 118 not upgraded.
```
Also already installed

**Comment by skmp** [comment]
can you `rm -rf` `Build/unittests/FEXLinuxTests/`?

**Comment by Sonicadvance1** [comment]
> unittests/FEXLinuxTests/

done

**Comment by skmp** [comment]
Should be resolved with ebe7c2f5

**Comment by Sonicadvance1** [comment]
Before merging I need to run this through pbuilder to ensure it works there as well

**Comment by skmp** [comment]
@Sonicadvance1 looks like I finally got this to behave -.-

File: unittests/FEXLinuxTests/CMakeLists.txt [hunk 6]
```diff
@@ -1,31 +1,33 @@
...
-  INSTALL_COMMAND ""
-  BUILD_ALWAYS ON
-  )
+set(GuestArchList "x86_32;x86_64")
+
+foreach(GUEST_ARCH ${GuestArchList})
+  set(TOOLCHAIN_FILE ${TOOLCHAIN_FILE_${GUEST_ARCH}})
+
+  # make sure configure is run again if the toolchain file is modified
+  set_directory_properties(PROPERTIES CMAKE_CONFIGURE_DEPENDS "${TOOLCHAIN_FILE}")
```
Comment (neobrain) [inline unittests/FEXLinuxTests/CMakeLists.txt:9]: AFAICT this won't catch modifications to the `-generic` toolchain file.

File: Toolchains/toolchain-x86_64.cmake [hunk 7]
```diff
@@ -0,0 +1,2 @@
+set(GUEST_ARCH "x86_32")
```
Comment (Sonicadvance1) [inline Toolchains/toolchain-x86_64.cmake:1]: Typo. This should be x86_64.

File: .github/workflows/ccpp.yml [hunk 8]
```diff
@@ -26,6 +26,14 @@ jobs:
     steps:
     - uses: actions/checkout@v2
 
+    - name: Environment info
```
Comment (Sonicadvance1) [inline .github/workflows/ccpp.yml:29]: This has no reason to exist anymore.

**Comment by Sonicadvance1** [comment]
Testing this on HDK888, every thunk usage results in a crash.

File: .github/workflows/ccpp.yml (hunk 8 above)
Comment (skmp) [reply .github/workflows/ccpp.yml:29]: It’s good for debug though, isn’t it? Don’t have to remember runner conf

Comment (Sonicadvance1) [reply .github/workflows/ccpp.yml:29]: Sure.

**Comment by Sonicadvance1** [comment]
```
(gdb) bt
#0  0x0000000000001401 in ?? ()
#1  0x0000007fe0a381ac in Invoke<void, unsigned int, int, int, int, int, int, unsigned int, unsigned int, void const*, unsigned long> (func=0x1401, args=...)
    at /mnt/Work/Work/work/FEXNew/ThunkLibs/HostLibs/../include/common/PackedArguments.h:130
#2  CallbackUnpack<void (unsigned int, int, int, int, int, int, unsigned int, unsigned int, void const*)>::ForIndirectCall(void*) (argsv=<optimized out>)
    at /mnt/Work/Work/work/FEXNew/ThunkLibs/HostLibs/../include/common/Host.h:176
#3  0x0000007f83862cf4 in ?? ()
```

File: unittests/FEXLinuxTests/CMakeLists.txt (hunk 6 above)
Comment (skmp) [reply unittests/FEXLinuxTests/CMakeLists.txt:9]: seriously? doh cmake.

**Comment by skmp** [comment]
@Sonicadvance1 can you add `[libc6-dev-amd64-cross](https://packages.ubuntu.com/jammy/libc6-dev-amd64-cross)` to the x86_64 runner?

File: unittests/FEXLinuxTests/CMakeLists.txt (hunk 6 above)
Comment (skmp) [reply unittests/FEXLinuxTests/CMakeLists.txt:9]: Fixed

**Comment by Sonicadvance1** [comment]
```
ryanh@buildbot-NUC10i7FNH:~$ sudo apt-get install linux-libc-dev-amd64-cross
[sudo] password for ryanh:
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  linux-libc-dev-amd64-cross
0 upgraded, 1 newly installed, 0 to remove and 22 not upgraded.
Need to get 1,240 kB of archives.
After this operation, 6,393 kB of additional disk space will be used.
Get:1 http://us.archive.ubuntu.com/ubuntu jammy/universe amd64 linux-libc-dev-amd64-cross all 5.15.0-22.22cross3 [1,240 kB]
Fetched 1,240 kB in 1s (1,903 kB/s)
Selecting previously unselected package linux-libc-dev-amd64-cross.
(Reading database ... 223389 files and directories currently installed.)
Preparing to unpack .../linux-libc-dev-amd64-cross_5.15.0-22.22cross3_all.deb ...
Unpacking linux-libc-dev-amd64-cross (5.15.0-22.22cross3) ...
Setting up linux-libc-dev-amd64-cross (5.15.0-22.22cross3) ...
```
Installed, reran CI.

**Comment by skmp** [comment]
the x86 runner also needed `libc6-dev-amd64-cross`

File: Toolchains/toolchain-x86_64.cmake (hunk 7 above)
Comment (skmp) [reply Toolchains/toolchain-x86_64.cmake]: fixed

**Comment by Sonicadvance1** [comment]
This doesn't immediately fault now at least. I'll need to poke around and see if there were any regressions

File: unittests/FEXLinuxTests/tests/CMakeLists.txt (hunk 2 above)
Comment (neobrain) [reply unittests/FEXLinuxTests/tests/CMakeLists.txt:67]: Commented out debug messages like this usually end up breaking due to some refactor not updating them. These aren't complicated fmt-style `print this entire container` logs that are hard to come up with, but simple variable prints that are easily on-demand if someone needs them.

**Review by neobrain — CHANGES_REQUESTED** [review]
Would prefer if these changes were organized in better-structured commits. I made an exception for reviewing since it's only a small diff overall.

File: unittests/FEXLinuxTests/CMakeLists.txt [hunk 9]
```diff
@@ -1,31 +1,33 @@
...
+  message("FEXLinuxTests: toolchain for |${GUEST_ARCH}| is \"${TOOLCHAIN_FILE}\"")
+  ExternalProject_Add(FEXLinuxTests-${GUEST_ARCH}
+    PREFIX FEXLinuxTests-${GUEST_ARCH}
+    SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}/tests"
+    BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}/FEXLinuxTests-${GUEST_ARCH}"
+    CMAKE_ARGS
+    "-DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}"
+    "-DCMAKE_TOOLCHAIN_FILE:FILEPATH=${TOOLCHAIN_FILE}"
+    "-DGUEST_ARCH=${GUEST_ARCH}"
+    "-DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}"
```
[the remark this answers is not in the document]
Comment (skmp) [reply unittests/FEXLinuxTests/CMakeLists.txt:20]: Because CMake is full of inconsistencies, and the toolchain is not always loaded before the cmake script runs.

File: Toolchains/toolchain-generic.cmake [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    # the triple identifies the target platform
+    set(triple x86_64-linux-gnu)
+    set(sysroot_triple x86_64-linux-gnu)
+else()
+    message(FATAL_ERROR "Toolchain doesn't support guest arch: |${GUEST_ARCH}|")
+endif()
+
+# Force re-checking of the compiler - cmake sometimes gets confused here and skips it
+set(CMAKE_C_ABI_COMPILED False)
```
[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:20]: if `CMAKE_C_ABI_COMPILED` is true, the compiler is not checked. CMake has a bug around toolchains there, and it gets confused that host clang == guest clang.

File: Toolchains/toolchain-generic.cmake [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+set(CMAKE_CXX_ABI_COMPILED False)
+
+
+# By default we use the clang & lld that was used to compile and link FEX, in cross compiler mode
+#
+set(CMAKE_C_COMPILER clang)
+set(CMAKE_CXX_COMPILER clang++)
+# we need to use lld for cross compiling
```
[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:28]: uh, because ld can't cross link?

[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:28]: Also, CMake does not support loading the arch specific linker, as it uses the compiler driver to invoke it

File: Toolchains/toolchain-generic.cmake [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+add_link_options("-fuse-ld=lld")
+
+
+# gcc alternative
+#
+## set(CMAKE_C_COMPILER i686-linux-gnu-gcc)
```
[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:34]: It is somewhat functional, yes, that is the old path we were using.

File: Toolchains/toolchain-generic.cmake [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+set(CMAKE_SYSROOT_LINK "/lib/${sysroot_triple}")
+add_link_options("-L/usr/${sysroot_triple}/lib" "-L/usr/lib/${sysroot_triple}")
+
+set(CMAKE_FIND_ROOT_PATH "/usr/${sysroot_triple}")
+
+
+set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
+set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
+set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
```
[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:57]: I'll see if my editor has an option about that, didn't know about that git shortcoming

File: Toolchains/toolchain-generic.cmake (hunk 11 above)
[the remark this answers is not in the document]
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:28]: > Why? Should be mentioned in the comment.

TBH, if anyone wants to understands how these shenanigans to get cmake toolchains to work, they should probably hit me up on discord, I'm not gonna write a 100-page cmake defect report for this.

**Comment by skmp** [comment]
> Would prefer if these changes were organized in better-structured commits. I made an exception for reviewing since it's only a small diff overall.

I can squash if you prefer, though that's often less useful. As we do have clean git log history as a non-goal in FEX, I don't think this is important though.

**Comment by neobrain** [comment]
The toolchain changes look good, but overall this gets a -1 from me due to the added build system->C++ source dependency in FEXLinuxTests.

Using a regexp on a source file to decide how to build that source file seems like asking for future trouble. We only have few uses of this pattern currently (all in FexLinuxTests), and I would like us to work to get rid of those instead of adding more.

File: Toolchains/toolchain-generic.cmake [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+## set(CMAKE_C_COMPILER i686-linux-gnu-gcc)
+## set(CMAKE_C_FLAGS -m32)
+## set(CMAKE_CXX_COMPILER i686-linux-gnu-g++)
+## set(CMAKE_CXX_FLAGS -m32)
+
+set(CMAKE_C_COMPILER_TARGET ${triple})
+set(CMAKE_CXX_COMPILER_TARGET ${triple})
+
+# Set the compiler to looks for libs, includes and packages in the correct folders
+# as seen on the official cmake site
```
Comment (neobrain) [inline Toolchains/toolchain-generic.cmake:43]: Can you provide a specific reference? I couldn't find documentation on the official cmake site that seemed relevant to what we're doing here.

File: Toolchains/toolchain-generic.cmake [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+set(CMAKE_SYSROOT_COMPILE "/usr/${sysroot_triple}")
+set(CMAKE_C_STANDARD_INCLUDE_DIRECTORIES "/usr/include")
```
Comment (neobrain) [inline Toolchains/toolchain-generic.cmake:45]: Why is this needed? Breaks the build on my non-Ubuntu x86 development machine, would be good to know what trade-offs I'm making by commenting it out.

Comment (skmp) [reply Toolchains/toolchain-generic.cmake:45]: non-Ubuntu builds are not supported at this time

File: Toolchains/toolchain-generic.cmake (hunk 14 above)
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:43]: https://cmake.org/cmake/help/latest/variable/CMAKE_SYSROOT_COMPILE.html

**Comment by skmp** [comment]
> The toolchain changes look good, but overall this gets a -1 from me due to the added build system->C++ source dependency in FEXLinuxTests.

Again, as stated before, that is how FLTs are designed  to work, and this is hardly something out of the line for this kind of testing, as one wants the test cases to be self-describing for beverity.

> Using a regexp on a source file to decide how to build that source file seems like asking for future trouble. 

Can you elaborate on why?

> If we can find a more structured way on adding the architecture metadata, 

The architecture metadata is added in a structured way, the same way the rest of the compile metadata is added. 

> I'm fine with merging this PR, though minor issues remain such as the leftover debug messages and the lack of documentation for nontrivial adjustments made in the toolchain.

I think this level of documentation is outside our scope right now. We don't document the maths behind SSA either, just like we don't include a copy of the x86 and arm64 architecture manuals.

File: Toolchains/toolchain-generic.cmake (hunk 15 above)
Comment (neobrain) [reply Toolchains/toolchain-generic.cmake:45]: I'm not saying this critically needs to change, but could you comment on why it's needed so I understand what consequences this has on the dev machine I actively use?

Comment (skmp) [reply Toolchains/toolchain-generic.cmake:45]: Well, an ubuntu dev machine should be used for FEX, one could say. Leaving that aside, a cross compile setup for inconsistently picks up the wrong include paths, so this is required.

Outside of ubuntu, only debian supports multiarch, afair, so you are in a world of pain using an unsupported distro. Likewise for rootfs (which needs to match your host os version and host library versions for predictable results).

File: Toolchains/toolchain-generic.cmake (hunk 14 above)
Comment (skmp) [reply Toolchains/toolchain-generic.cmake:43]: The defaults in cross compiling are actually empty, it seems, so you need to set it up. Sadly, random partially working examples and this documentation is all that the cmake team provides - google is your friend to find out more info.

I don't have the exact reference handy, it was like 2-3 google searches away when I found it and copy pasted it.

Comment (skmp) [reply Toolchains/toolchain-generic.cmake:43]: In general, I think the only 'predictable' way to support this is having a full dev rootfs, as we have discussed in the meets, which kind of falls outside the scope of this PR. @Sonicadvance1 said they will look into it when they have iirc.

**Comment by skmp** [comment]
@Sonicadvance1 can you go over on this and give a quick ACK/NACK?

It's a prerequisite for several other things I'm working on and having it out of main significantly slows things down.

As far as I understand you had some issue with FLTs with this - which doesn't happen in the latest iteration?

**Comment by skmp** [comment]
(Closing this as there is an ongoing powergrab by @Sonicadvance1, I will migrate my work to https://github.com/skmp/fex-emu-ng.git)
